### ISSUE TITLE
🔷 Prototype stateless validation

### ISSUE THREAD

**Issue by robin-near** [near/nearcore#9292]
```[tasklist]
### Tasks
- [x] Implement state proof and validation based on MPT
- [ ] https://github.com/near/nearcore/issues/9628
- [x] Estimate state witness size from historical chunks
- [ ] https://github.com/near/nearcore/issues/9982
- [ ] Implement state witness size limit
- [x] Investigate requirements for putting all state in memory
- [ ] https://github.com/near/nearcore/issues/9511
- [ ] Stateless validation stage 0 NEP
- [ ] https://github.com/near/nearcore/issues/10265
```

**Comment by walnut-the-cat** [comment]
Sept 14th
- Overall project timeline/roadmap is shared: [link](https://docs.google.com/document/d/1TzMENFGYjwc2g5A3Yf4zilvBwuYJufsUQJwRjXGb9Xc/edit?usp=sharing)
- Came up with several proposal to address storage performance issues before e2e implementation of stateless validation
- Moving forward with prototyping short term solution for now while evaluating feasibility/necessity of midterm solutions
- Requirement gathering for NEP is in progress

- Post state root:
  - prototyping in progress (No major progress this week due to Anton’s OOO). Alex will start contributing towards this effort

- In-memory trie:
  - Several options have been discussed to minimize additional RAM requirement for validators. Dynamic memory allocation for in-memory trie with mmap is being prototyped

**Comment by walnut-the-cat** [comment]
*[Note] The issue will stay open until stage 0 of stateless validation is complete*

Sep 28th
* Please refer to post state root project and in memory trie project for more details
* Synced up with Jakob to get knowledge transfer for limiting state witness size

Overall October roadmap is as follows:
* **Target end state**: Successful run of stateless validation MVP in localnet (without optimization)
* Post state root: [#9450](https://github.com/near/nearcore/issues/9450) 
  * Code complete to unblock e2e MVP implementation of stateless validation stage 0
    * Out of scope: optimization, full test coverage
* In memory trie: [#9511](https://github.com/near/nearcore/issues/9511)
  * Complete implementation of in-memory trie
  * partial in-memory trie ([ref](https://docs.google.com/document/d/1TzMENFGYjwc2g5A3Yf4zilvBwuYJufsUQJwRjXGb9Xc/edit#heading=h.g1xqke6gf3ys))
* State witness implementation for MVP ([ref](https://docs.google.com/document/d/1TzMENFGYjwc2g5A3Yf4zilvBwuYJufsUQJwRjXGb9Xc/edit?usp=sharing))
  * Generation by chunk proposer
  * Delivery over network
  * Witness validation by validator
  * Chunk endorsement by validator
  * Chunk endorsement collection by block producer
  * Out of scope: optimistic block production
* Finalize NEP draft for post state root and schedule for SME review: [ref](https://github.com/near/NEPs/pull/507/)
* Finalize NEP draft for stateless validation stage 0: [ref](https://github.com/near/NEPs/pull/509/)

Network related
* **Target end state**: Robust plans for stateless validation related network projects and execution start
  * [Stretch] Complete implementation
* Adding new message types for stateless validation
* Data dependency bomb, including slow distribution of large chinks ([ref](https://github.com/near/nearcore-private/issues/93))
* Limiting state witness size ([ref](https://github.com/near/nearcore/issues/9378))

**Comment by Longarithm** [comment]
Completed couple months ago, see https://github.com/near/near-one-project-tracking/issues/5 for further progress.
